plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotDataFetcher.java

```diff
@@ -0,0 +1,89 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.pinot.client;
+
+import io.trino.plugin.pinot.PinotException;
+import io.trino.plugin.pinot.PinotSplit;
+import io.trino.spi.connector.ConnectorSession;
+import org.apache.pinot.common.utils.DataTable;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static io.trino.plugin.pinot.PinotErrorCode.PINOT_EXCEPTION;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public interface PinotDataFetcher
+{
+    default void checkExceptions(DataTable dataTable, PinotSplit split, String query)
+    {
+        Map<String, String> metadata = dataTable.getMetadata();
+        List<String> exceptions = new ArrayList<>();
+        metadata.forEach((k, v) -> {
+            if (k.startsWith(DataTable.EXCEPTION_METADATA_KEY)) {
+                exceptions.add(v);
+            }
+        });
+        if (!exceptions.isEmpty()) {
+            throw new PinotException(PINOT_EXCEPTION, Optional.of(query), format("Encountered %d pinot exceptions for split %s: %s", exceptions.size(), split, exceptions));
+        }
+    }
+
+    long getReadTimeNanos();
+
+    long getMemoryUsageBytes();
+
+    boolean endOfData();
+
+    boolean isDataFetched();
+
+    void fetchData();
+
+    PinotDataTableWithSize getNextDataTable();
+
+    class RowCountChecker
+    {
+        private final AtomicLong currentRowCount = new AtomicLong();
+        private final int limit;
+        private final String query;
+
+        public RowCountChecker(int limit, String query)
+        {
+            this.limit = limit;
+            this.query = requireNonNull(query, "query is null");
+        }
+
+        public void checkTooManyRows(DataTable dataTable)
+        {
+            if (currentRowCount.addAndGet(dataTable.getNumberOfRows()) > limit) {
+                throw new PinotException(PINOT_EXCEPTION, Optional.of(query), format("Segment query returned '%s' rows per split, maximum allowed is '%s' rows.", currentRowCount.get(), limit));
+            }
+        }
+    }
+
+    interface Factory
+    {
+        PinotDataFetcher create(ConnectorSession session, String query, PinotSplit split);
+    }
+
+    interface PinotServerQueryClient
+    {
+        Iterator<PinotDataTableWithSize> queryPinot(ConnectorSession session, String query, String serverHost, List<String> segments);
```

Comment:
Why do we need both a Factory and a PinotServerQueryClient interface? I'm not sure I understand the relationship between the two. Why do we need different implementations of the Factory, given that the query client is already abstracted to support both underlying protocols?

Reply:
Make sense, I think we just need factory no need for PinotServerQueryClient

Reply:
Sounds good! Updating.